### PR TITLE
Add Recaptcha functionality to Gitea

### PR DESCRIPTION
**Pull request by ghost**
This PR adds recaptcha functionality to Gitea.

Old captcha functionality remains the same, this just gives options to the user
recaptcha

### CODE REVIEW

**Review by appleboy — APPROVED** [review]

File: custom/conf/app.ini.sample [hunk 1]
```diff
@@ -299,7 +299,12 @@ ENABLE_NOTIFY_MAIL = false
 ENABLE_REVERSE_PROXY_AUTHENTICATION = false
 ENABLE_REVERSE_PROXY_AUTO_REGISTRATION = false
 ; Enable captcha validation for registration
-ENABLE_CAPTCHA = true
+ENABLE_CAPTCHA = false
```
Comment (lafriks) [inline custom/conf/app.ini.sample:304]: Do not change default as this way it will be breaking change

**Review by lafriks — CHANGES_REQUESTED** [review]

File: modules/templates/helper.go [hunk 2]
```diff
@@ -186,6 +186,12 @@ func NewFuncMap() []template.FuncMap {
 		"ParseDeadline": func(deadline string) []string {
 			return strings.Split(deadline, "|")
 		},
+		"EnableRecaptcha": func() bool {
```
Comment (jonasfranz) [inline modules/templates/helper.go]: Why do you need a helper function for that? Please use `ctx.Data`.

File: routers/user/auth.go [hunk 3]
```diff
@@ -760,6 +767,16 @@ func LinkAccountPostRegister(ctx *context.Context, cpt *captcha.Captcha, form au
 		return
 	}
 
+	if setting.Service.EnableRecaptcha {
+		ctx.Req.ParseForm()
+		valid, _ := recaptcha.Verify(ctx.Req.Form.Get("g-recaptcha-response"))
```
Comment (jonasfranz) [inline routers/user/auth.go]: Add `g-recaptcha-response` as form field to `form`.

File: routers/user/auth.go [hunk 4]
```diff
@@ -887,6 +910,16 @@ func SignUpPost(ctx *context.Context, cpt *captcha.Captcha, form auth.RegisterFo
...
```
Comment (jonasfranz) [inline routers/user/auth.go]: Add `g-recaptcha-response` as form field to `auth.RegisterForm`.

File: routers/user/auth_openid.go [hunk 5]
```diff
@@ -341,6 +346,16 @@ func RegisterOpenIDPost(ctx *context.Context, cpt *captcha.Captcha, form auth.Si
 		return
 	}
 
+	if setting.Service.EnableRecaptcha {
+		ctx.Req.ParseForm()
+		valid, _ := recaptcha.Verify(ctx.Req.Form.Get("g-recaptcha-response"))
```
Comment (jonasfranz) [inline routers/user/auth_openid.go]: Same as above.

File: templates/user/auth/signup_inner.tmpl [hunk 6]
```diff
@@ -40,6 +40,13 @@
 							</div>
 						{{end}}
 
+						{{if .EnableRecaptcha}}
+							<div class="inline field required">
+								<div class="g-recaptcha" data-sitekey="{{ .RecaptchaSitekey }}" style="margin: 0 auto; width: 304px; padding-left: 30px"></div>
```
Comment (jonasfranz) [inline templates/user/auth/signup_inner.tmpl]: Do not use inline css. Please add the style to the `.less` files.

File: templates/user/auth/signup_inner.tmpl [hunk 7]
```diff
@@ -40,6 +40,13 @@
...
+								<script src="https://www.google.com/recaptcha/api.js" async></script>
```
Comment (jonasfranz) [inline templates/user/auth/signup_inner.tmpl]: Scripts should be placed in footer.

File: templates/user/auth/signup_openid_register.tmpl [hunk 8]
```diff
@@ -30,6 +30,12 @@
 							<input id="captcha" name="captcha" value="{{.captcha}}" autocomplete="off">
 						</div>
 					{{end}}
+					{{if .EnableRecaptcha}}
+						<div class="inline field required">
+							<div class="g-recaptcha" data-sitekey="{{ .RecaptchaSitekey }}" style="margin: 0 auto; width: 304px; padding-left: 30px"></div>
+							<script src="https://www.google.com/recaptcha/api.js" async></script>
```
Comment (jonasfranz) [inline templates/user/auth/signup_openid_register.tmpl]: Same as above.

File: modules/recaptcha/recaptcha.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+		return false, fmt.Errorf("Failed to read CAPTCHA response: %s", err)
+	}
+	jsonResponse := Response{Success: false} // set a default of fail for CAPTCHA
+	err = json.Unmarshal(body, &jsonResponse)
+	if err != nil {
+		return false, fmt.Errorf("Failed to parse CAPTCHA response: %s", err)
+	}
+
+	return jsonResponse.Success, nil
+
```
Comment (jonasfranz) [inline modules/recaptcha/recaptcha.go]: Remove empty line

File: modules/recaptcha/recaptcha.go [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+		url.Values{"secret": {setting.Service.RecaptchaSecret}, "response": {response}})
+	if err != nil {
+		return false, fmt.Errorf("Failed to send CAPTCHA response: %s", err)
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return false, fmt.Errorf("Failed to read CAPTCHA response: %s", err)
+	}
+	jsonResponse := Response{Success: false} // set a default of fail for CAPTCHA
```
Comment (jonasfranz) [inline modules/recaptcha/recaptcha.go]: The default value of `Success` is `false` => No need for initializing it.

**Review by jonasfranz — CHANGES_REQUESTED** [review]

File: custom/conf/app.ini.sample (hunk 1 above)
Comment (jonasfranz) [reply custom/conf/app.ini.sample:304]: The default value is actually `false` but the documentation is wrong

Comment (ghost) [reply custom/conf/app.ini.sample:304]: Hi. Yes `false` is the default, and I changed it to be correct like @JonasFranzDEV said.

**Comment by ghost** [comment]
Hi @lafriks and @JonasFranzDEV I made changes as requested. When you have chance please review. Thank you.

**Comment by jonasfranz** [comment]
Please run `make generate-stylesheets` and commit the result.
The position of the reCAPTCHA is not inline with the other fields. And on mobile devices it overlaps.
bildschirmfoto vom 2018-05-25 um 21 26 34
screenshot-2018-5-25 gitea git with a cup of tea

**Comment by ghost** [comment]
Thank you @JonasFranzDEV I've updated the CSS ta make it responsive for recaptcha. Please see the following screenshots:

recaptchabig
recaptchasmall

**Comment by cezar97** [comment]
@flufmonster Why not use the invisible one to maximize user experience? This way the capthcha doesn't have to be solved on each login but only if detected as bot.

**Comment by ghost** [comment]
Thank you for feedback @cezar97. This only shows on sign up page, not on login.

**Review by jonasfranz — APPROVED** [review]
LGTM, but this PR could just be merged after the 1.5.0 release because we're in feature freeze currently. It also requires the approval of @lafriks .

**Comment by axifive** [comment]
Maybe need to use [compact size](https://www.google.com/recaptcha/api2/demo?size=compact) option for mobile view?

**Review by lafriks — APPROVED** [review]

**Comment by techknowlogick** [comment]
@flufmonster please resolve conflict.

File: custom/conf/app.ini.sample (hunk 1 above)
Comment (lunny) [reply custom/conf/app.ini.sample:304]: I would like you use a captcha type and currently you have three options: disabled, image and recaptcha and default is disabled.

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
@lunny to keep backwards compatibility for people who have `ENABLE_CAPTCHA=true` I add another option to change type of recaptcha instead of being able to have both at same time.

File: routers/user/auth.go [hunk 11]
```diff
@@ -755,12 +762,21 @@ func LinkAccountPostRegister(ctx *context.Context, cpt *captcha.Captcha, form au
 		return
 	}
 
-	if setting.Service.EnableCaptcha && !cpt.VerifyReq(ctx.Req) {
+	if setting.Service.EnableCaptcha && setting.Service.CaptchaType == "image" && !cpt.VerifyReq(ctx.Req) {
```
Comment (lunny) [inline routers/user/auth.go]: Could you save "image" and "recaptcha" as a const strings and only reference the const variables?

**Review by lunny — CHANGES_REQUESTED** [review]

File: routers/user/auth.go (hunk 11 above)
Comment (ghost) [reply routers/user/auth.go]: ok. I changed this.

**Review by lunny — APPROVED** [review]

**Pull request merged**